protocol-select/README.md

```diff
@@ -0,0 +1,541 @@
+
+Non-breaking changes to the protocol can be done at the schema level, more
+specifically through the _Protocol Buffer_ framework. Instead of enumerating the
+various update mechanisms, we refer to the _[Updating a Message Type]_ section
+of the _Protocol Buffer_ specification.
+
+As an example for a non-breaking change, say we would like to exchange a made up
+name via the _Protocol Select_ protocol. We can simply extend the `ProtoSelect`
+message type by an `optional string name = 4;` field. Updated implementations
+would be able to extract the name from the payload, old implementations would
+simply ignore the new field.
+
+#### Breaking Changes
+
+When making breaking changes to the _Protocol Select_ protocol,
+implementations need to be able to differentiate the old and the new version on
+the wire. This is done via the `version` field in the `ProtoSelect` message,
+treated as an ordinal monotonically increasing number, with each increase
+identifying a new breaking version of the protocol.
+
+As an example for a made-up breaking change, say we would like the listed
+protocols in the `ProtoSelect` message to enumerate the protocols that the local node
+does *not* support. One would bump the `version` field by `1`. Implementations
+supporting both versions are able to differentiate an old and new version
+message. Implementations supporting only the old version would reject a new
+version message and fail the negotiation. Roll-out strategies need to cope with
+such negotiation failure, e.g. through retries with an older version.
+
+
+## Extensions
+
+### Protocol IDs
+
+The first version of _Protocol Select_ will allow specifying protocols by their
+_Protocol Name_, i.e. human readable string representation, only. In order to
+optimize on bandwidth, future versions might introduce alternative
+representations in a non-breaking manner.
+
+More specifically, this extension would allow specifying protocols by their
+_Protocol ID_. A _Protocol ID_ is a [Multicodec] or a combination of
+[Multicodec]s. Implementations can specify a protocol either via a _Protocol
+Name_ or a _Protocol ID_ by extending the `Protocol` message type definition as
+follows:
+
+```diff
+message Protocol {
+    oneof protocol {
+        string name = 1;
++       uint64 id = 2;
+    }
+}
+```
+
+_Protocol Name_ and _Protocol ID_ can be used interchangeably. To ease roll-out
+of a _Protocol ID_ for a protocol that has previously been negotiated via its
+_Protocol Name_, one might leverage one (or multiple) of the following
+mechanisms:
+
+- Extending the libp2p identify protocol, allowing nodes to announce their
+  supported protocols both by _Protocol Name_ and _Protocol ID_, thus signaling
+  the support for the _Protocol ID_ extension for the concrete protocols.
+
+- Including a protocol both by its _Protocol Name_ and _Protocol ID_ in the list
+  of supported protocols.
+
+  Note, when optimistically negotiating a stream protocol as an initiator, with a
```

Comment:
Let's note _why_ one might want this. Specifically, you'll _learn_ that they support the protocol ID.

Also note: Figuring out how to implement this internally will be... a bit interesting. Currently we just listen on a protocol. Now, we'll need some way to specify the mapping.

Reply:
Noted in 843e76097a50f7bad3693f31e24fe64031f0a865.